source/grid/grid_tools.cc

```diff
@@ -1606,6 +1606,45 @@ namespace GridTools
   }
 
 
+  template <class MeshType>
+  std::vector<typename MeshType::cell_iterator>
+  compute_cell_halo_layer_on_level
```

Comment:
This was the only change I was a bit unsure on. I added the function <tt>compute_cell_halo_layer_on_level()</tt> as an internal function in <tt>shared_tria.cc</tt> directly above the function <tt>partition()</tt>. Let me know if I have misunderstood you and you had something else in mind.